File: src/lib/wow/archive/casc/tact-keys.ts
/**
 * TACT decryption key management, ported from wow.export (src/js/casc/tact-keys.js).
 */
import { promises as fsp } from 'fs';
import path from 'path';

import { write } from '@/lib/wow/log';

import { constants } from '../../formats/constants';
import { get } from '../../formats/generics';
import { wowConfig } from '../../server/config';

const KEY_RING: Record<string, string> = {};
let isSaving = false;

/** Retrieve a registered decryption key. */
export function getKey(keyName: string): string | undefined {
  return KEY_RING[keyName.toLowerCase()];
}

/** Validate a keyName/key pair. */
function validateKeyPair(keyName: string, key: string): boolean {
  if (keyName.length !== 16) return false;
  if (key.length !== 32) return false;
  return true;
}

/**
 * Add a decryption key. Subject to validation.
 * Decryption keys will be saved to disk on next tick.
 * Returns true if added, else false if the pair failed validation.
 */
export function addKey(keyName: string, key: string): boolean {
  if (!validateKeyPair(keyName, key)) return false;

  const normalizedName = keyName.toLowerCase();
  const normalizedKey = key.toLowerCase();

  if (KEY_RING[normalizedName] !== normalizedKey) {
    KEY_RING[normalizedName] = normalizedKey;
    write('Registered new decryption key %s -> %s', normalizedName, normalizedKey);
    void save();
  }

  return true;
}

/**
 * Load tact keys from disk cache and request updated keys from remote server.
 */
export async function load(): Promise<void> {
  // Load from local cache.
  try {
    const tactKeys = JSON.parse(await fsp.readFile(constants.CACHE.TACT_KEYS, 'utf8')) as Record<string, string>;

    // Validate/add our cached keys manually rather than passing to addKey()
    // to skip over redundant logging/saving calls.
    let added = 0;
    for (const [keyName, key] of Object.entries(tactKeys)) {
      if (validateKeyPair(keyName, key)) {
        KEY_RING[keyName.toLowerCase()] = key.toLowerCase();
        added++;
      } else {
        write('Skipping invalid tact key from cache: %s -> %s', keyName, key);
      }
    }

    write('Loaded %d tact keys from local cache.', added);
  } catch (e) {
    // No tactKeys cached locally, doesn't matter.
  }

  const tactUrl = wowConfig.tactKeysURL;
  const tactUrlFallback = wowConfig.tactKeysFallbackURL;
  const res = await get([tactUrl, tactUrlFallback]);

  if (!res.ok) throw new Error(`Unable to update tactKeys, HTTP ${res.status}`);

  const data = await res.text();
  const lines = data.split(/\r\n|\n|\r/);
  let remoteAdded = 0;

  for (const line of lines) {
    const parts = line.split(' ');
    if (parts.length !== 2) continue;

    const keyName = parts[0].trim();
    const key = parts[1].trim();

    if (validateKeyPair(keyName, key)) {
      KEY_RING[keyName.toLowerCase()] = key.toLowerCase();
      remoteAdded++;
    } else {
      write('Skipping invalid remote tact key: %s -> %s', keyName, key);
    }
  }

  if (remoteAdded > 0) write('Added %d tact keys from %s', remoteAdded, tactUrl);
}

/**
 * Request for tact keys to be saved on the next tick.
 * Multiple calls can be chained in the same tick.
 */
function save(): void {
  if (!isSaving) {
    isSaving = true;
    setImmediate(() => void doSave());
  }
}

/** Saves the tact keys to disk. */
async function doSave(): Promise<void> {
  await fsp.mkdir(path.dirname(constants.CACHE.TACT_KEYS), { recursive: true });
  await fsp.writeFile(constants.CACHE.TACT_KEYS, JSON.stringify(KEY_RING, null, '\t'), 'utf8');
  isSaving = false;
}

export default { load, getKey, addKey };

File: golang/internal/wow/archive/casc/tact_keys.go
package casc

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/wow/config"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var (
	keyRing  = map[string]string{}
	keyMu    sync.RWMutex
	isSaving bool
	saveMu   sync.Mutex
)

// GetKey retrieves a registered decryption key.
func GetKey(keyName string) (string, bool) {
	keyMu.RLock()
	defer keyMu.RUnlock()
	key, ok := keyRing[strings.ToLower(keyName)]
	return key, ok
}

func validateKeyPair(keyName, key string) bool {
	return len(keyName) == 16 && len(key) == 32
}

// AddKey adds a decryption key subject to validation.
func AddKey(keyName, key string) bool {
	if !validateKeyPair(keyName, key) {
		return false
	}
	normalizedName := strings.ToLower(keyName)
	normalizedKey := strings.ToLower(key)
	keyMu.Lock()
	defer keyMu.Unlock()
	if keyRing[normalizedName] != normalizedKey {
		keyRing[normalizedName] = normalizedKey
		log.Write("Registered new decryption key %s -> %s", normalizedName, normalizedKey)
		scheduleSave()
	}
	return true
}

// LoadTactKeys loads tact keys from disk cache and remote servers.
func LoadTactKeys() error {
	if data, err := os.ReadFile(constants.Cache.TactKeys); err == nil {
		var tactKeys map[string]string
		if json.Unmarshal(data, &tactKeys) == nil {
			added := 0
			keyMu.Lock()
			for keyName, key := range tactKeys {
				if validateKeyPair(keyName, key) {
					keyRing[strings.ToLower(keyName)] = strings.ToLower(key)
					added++
				} else {
					log.Write("Skipping invalid tact key from cache: %s -> %s", keyName, key)
				}
			}
			keyMu.Unlock()
			log.Write("Loaded %d tact keys from local cache.", added)
		}
	}

	remoteURLs := []string{config.WowConfig.TactKeysURL, config.WowConfig.TactKeysFallbackURL}
	for _, url := range remoteURLs {
		if url == "" {
			continue
		}
		res, err := formats.Get(url)
		if err != nil {
			log.Write("Failed to load tact keys from %s: %v", url, err)
			continue
		}
		if !res.OK {
			log.Write("Unable to update tactKeys from %s, HTTP %d", url, res.Status)
			continue
		}
		remoteAdded := mergeTactKeyLines(string(res.Body))
		if remoteAdded > 0 {
			log.Write("Added %d tact keys from %s", remoteAdded, url)
			scheduleSave()
		}
	}
	return nil
}

func mergeTactKeyLines(body string) int {
	lines := strings.FieldsFunc(body, func(r rune) bool { return r == '\r' || r == '\n' })
	remoteAdded := 0
	keyMu.Lock()
	for _, line := range lines {
		parts := strings.Split(line, " ")
		if len(parts) != 2 {
			continue
		}
		keyName := strings.TrimSpace(parts[0])
		key := strings.TrimSpace(parts[1])
		if validateKeyPair(keyName, key) {
			normalizedName := strings.ToLower(keyName)
			normalizedKey := strings.ToLower(key)
			if keyRing[normalizedName] != normalizedKey {
				keyRing[normalizedName] = normalizedKey
				remoteAdded++
			}
		} else {
			log.Write("Skipping invalid remote tact key: %s -> %s", keyName, key)
		}
	}
	keyMu.Unlock()
	return remoteAdded
}

func scheduleSave() {
	saveMu.Lock()
	defer saveMu.Unlock()
	if !isSaving {
		isSaving = true
		go func() {
			doSave()
			saveMu.Lock()
			isSaving = false
			saveMu.Unlock()
		}()
	}
}

func doSave() {
	keyMu.RLock()
	snapshot := make(map[string]string, len(keyRing))
	for k, v := range keyRing {
		snapshot[k] = v
	}
	keyMu.RUnlock()
	_ = os.MkdirAll(filepath.Dir(constants.Cache.TactKeys), 0o755)
	data, _ := json.MarshalIndent(snapshot, "", "\t")
	_ = os.WriteFile(constants.Cache.TactKeys, data, 0o644)
}
